File: python/huffmanEncoding.py
# lossless data compression algorithm 
# assigns variable-length codes to input characters, w/ the code length inversely proportional to the character's frequency
# more freqeuent characters get shorter codes and less frequent characters get longer codes 

from heapq import heappush, heappop 

class Node:
  def __init__(self, char, freq):
    self.char = char
    self.freq = freq
    self.left = None
    self.right = None

  def __lt__(self, other):
    return self.freq < other.freq 


def huffman_encoding(text):
  freq = {}
  for char in text:
    if char in freq:
      freq[char] += 1
    else:
      freq[char] = 1

  pq = []
  for char, freq in freq.items():
    heappush(pq, Node(char, freq))

  while len(pq) > 1:
    left = heappop(pq)
    right = heappop(pq)
    parent = Node(None, left.freq + right.freq)
    parent.left = left
    parent.right = right
    heappush(pq, parent)

  root = pq[0]

  code = {}
  def encode(node, curr_code):
    if node is None:
      return 
    if node.char is not None:
      code[node.char] = curr_code
      return
    encode(node.left, curr_code + "0")
    encode(node.right, curr_code + "1")

  encode(root, "")
  encoded_text = ""
  for char in text:
    encoded_text += code[char]
  return encoded_text, root 

def huffman_decoding(encoded_text, root):
  curr_node = root
  decoded_text = ""
  for bit in encoded_text:
    if bit == '0':
      curr_node = curr_node.left
    else:
      curr_node = curr_node.right
    if curr_node.char is not None:
      decoded_text += curr_node.char
      curr_node = root
  return decoded text 

if __name__ == "__main__":
  text = "Huffman coding is a data compression algorithm"
  encoded_text, root = huffman_encoding(text)
  print("encoded text:", encoded_text)

  decoded_text = huffman_decoding(encoded_text, root)
  print("decoded text:", decoded_text) 

File: javascript/huffmanEncoding.js
// lossless data compression algorithm
// assigns variable-length codes to input characters with the goal of minimizing the overall code length

class Node
  {
    constructor(data, freq)
    {
      this.data = data;
      this.freq = freq;
      this.left = null;
      this.right = null;
    }
  }
function huffmanEncoding(text)
{
  // create frequency array to store character frequencies
  const freq = [];
  for(let i = 0; i < 256; i++)
    {
      freq[i] = 0;
    }
  for(let i = 0; i < text.length; i++)
    {
      freq[text.charCodeAt(i)]++;
    }
  // create priority queue to store nodes
  const pq = [];
  for(let i = 0; i < 256; i++)
    {
      if(freq[i] != 0)
      {
        pq.push(new Node(i, freq[i]));
      }
    }
  pq.sort((a,b) => a.freq - b.freq);
  // build huffman tree
  while(pq.length > 1)
    {
      const x = pq.shift();
      const y = pq.shift();

      const node = new Node(null, x.freq + y.freq);
      node.left = x;
      node.right = y;

      pq.push(node);
      pq.sort((a, b) => a.freq - b.freq);
    }
  // print huffman codes
  const huffmanCode = [];
  const code = [];
  encode(pq[0], huffmanCode, code);

  for(let i = 0; i < 256; i++)
    {
      if(freq[i] != 0)
      {
        console.log('${chr(i)} ${huffmanCode[i]}');
      }
    }
}
function encode(root, huffmanCode, code)
{
  if(root == null)
    return;

  // assign 0 to left edge and right edge
  if(root.data != null)
  {
    huffmanCode[root.data] = code.join(' ');
  }
  code.push('0');
  encode(root.left, huffmanCode, code);
  code.pop();

  code.push('1');
  encode(root.right, huffmanCode, cdoe);
  code.pop();
}
// helper function to convert int to char
function chr(charCode)
{
  return String.fromCharCode(charCode);
}
// example usage
const text = "Huffman coding is a data compression algorithm";
huffmanEncoding(text); 
